Auto Migrate State into Terraform Cloud

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Description

<!--- Please leave a helpful description of the feature request here. --->

<!--- Information about code formatting: https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code --->

We're aiming to improve the usability for Terraform Cloud scenarios in cdktf. One part of this is the state migration into the Terraform Cloud. There are two scenarios:

1. Fresh / empty remote workspace
2. Remote workspace with pre-existing state

While it seems safe to assume that it's ok to auto-migrate state in the first case, we need to confirm with the user in the second case. 

The Terraform CLI is confirming even in he first case, and erroring out in the 2nd case.

We'll need to think about the full process and potential edge cases, if we really want to do this automatically - even when confirming with the user.

#### Terraform CLI 

##### No State

![Screenshot 2021-01-13 at 15 34 29](https://user-images.githubusercontent.com/136789/104467489-a7a5c580-55b6-11eb-911b-d4bf5be9f90b.png)

#####  Existing State

![Screenshot 2021-01-13 at 15 25 41](https://user-images.githubusercontent.com/136789/104467542-b68c7800-55b6-11eb-8c04-6df81a569285.png)

### References

<!---
Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests

Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor blog posts or documentation?
--->

- #449

This currently fails with 
```
[2023-01-11T16:24:01.625] [ERROR] default - ╷
│ Error: Can't ask approval for state migration when interactive input is disabled.
│
│ Please remove the "-input=false" option and try again.
│
│
```

We would need to run init in interactive mode as well to make it work

--- a/packages/@cdktf/cli-core/src/lib/cdktf-project.ts
+++ b/packages/@cdktf/cli-core/src/lib/cdktf-project.ts
@@ -75,6 +75,7 @@ export type DiffOptions = SingleStackOptions & {
   vars?: string[];
   varFiles?: string[];
   noColor?: boolean;
+  migrateState?: boolean;
 };
 
 export type MutationOptions = MultipleStackOptions &
@@ -86,6 +87,7 @@ export type MutationOptions = MultipleStackOptions &
     vars?: string[];
     varFiles?: string[];
     noColor?: boolean;
+    migrateState?: boolean;
   };
 
 export type LogMessage = {

--- a/packages/@cdktf/cli-core/src/lib/cdktf-stack.ts
+++ b/packages/@cdktf/cli-core/src/lib/cdktf-stack.ts
@@ -114,6 +114,7 @@ type CdktfStackOptions = {
   ) => void;
   onLog?: (log: { message: string; isError: boolean }) => void;
   autoApprove?: boolean;
+  migrateState?: boolean;
   abortSignal: AbortSignal;
 };
 
@@ -240,7 +241,11 @@ export class CdktfStack {
     }
 
     const needsUpgrade = await this.checkNeedsUpgrade();
-    await terraform.init(needsUpgrade, noColor);
+    await terraform.init(
+      needsUpgrade,
+      noColor ?? false,
+      this.options.migrateState ?? false
+    );
     return terraform;
   }
 
@@ -256,8 +261,12 @@ export class CdktfStack {
   }
 
   private async checkNeedsInit(): Promise<boolean> {
-    const lock = new TerraformProviderLock(this.stack.workingDirectory);
+    if (this.options.migrateState) {
+      // If we're migrating state, we need to init
+      return true;
+    }
 
+    const lock = new TerraformProviderLock(this.stack.workingDirectory);
     const lockFileExists = await lock.hasProviderLockFile();
 
     if (!lockFileExists) {

--- a/packages/@cdktf/cli-core/src/lib/models/deploy-machine.ts
+++ b/packages/@cdktf/cli-core/src/lib/models/deploy-machine.ts
@@ -1,6 +1,3 @@
-import * as os from "os";
-import * as path from "path";
-import * as fs from "fs";
 import {
   createMachine,
   send,
@@ -14,6 +11,7 @@ import * as pty from "@cdktf/node-pty-prebuilt-multiarch";
 import { Errors, logger } from "@cdktf/commons";
 import { missingVariable } from "../errors";
 import stripAnsi from "strip-ansi";
+import { spawnPty } from "./pty-process";
 
 interface PtySpawnConfig {
   file: Parameters<typeof pty.spawn>[0];
@@ -32,8 +30,8 @@ interface DeployContext {
 export type DeployEvent =
   | { type: "START"; pty: PtySpawnConfig }
   | { type: "STOP" }
-  | { type: "SEND_INPUT"; input: string }
-  | { type: "LINE_RECEIVED"; line: string } //TODO: rename to output received as not really a line in practice
+  | { type: "SEND_LINE"; input: string }
+  | { type: "OUTPUT_RECEIVED"; output: string }
   | { type: "APPROVED_EXTERNALLY" } // e.g. via TFC UI or API
   | { type: "REJECTED_EXTERNALLY" }
   | { type: "OVERRIDDEN_EXTERNALLY" }
@@ -97,9 +95,9 @@ export function extractVariableNameFromPrompt(line: string) {
 }
 
 export function handleLineReceived(send: (event: DeployEvent) => void) {
-  return (line: string) => {
-    let hideLine = false;
-    const noColorLine = stripAnsi(line);
+  return (output: string) => {
+    let hideOutput = false;
+    const noColorLine = stripAnsi(output);
 
     // possible events based on line
     if (noColorLine.includes("approved using the UI or API")) {
@@ -118,12 +116,12 @@ export function handleLineReceived(send: (event: DeployEvent) => void) {
       noColorLine.includes("var.") &&
       noColorLine.includes("Enter a value:")
     ) {
-      hideLine = true;
+      hideOutput = true;
 
-      const variableName = extractVariableNameFromPrompt(line);
+      const variableName = extractVariableNameFromPrompt(output);
       send({
-        type: "LINE_RECEIVED",
-        line: missingVariable(variableName),
+        type: "OUTPUT_RECEIVED",
+        output: missingVariable(variableName),
       });
       send({ type: "VARIABLE_MISSING", variableName });
     } else if (
@@ -131,18 +129,18 @@ export function handleLineReceived(send: (event: DeployEvent) => void) {
         "Do you want to override the soft failed policy check?"
       )
     ) {
-      hideLine = true;
-      send({ type: "LINE_RECEIVED", line });
+      hideOutput = true;
+      send({ type: "OUTPUT_RECEIVED", output });
 
       send({ type: "REQUEST_SENTINEL_OVERRIDE" });
     } else if (noColorLine.includes("overridden using the UI or API")) {
       send({ type: "OVERRIDDEN_EXTERNALLY" });
     }
 
-    if (!hideLine) {
+    if (!hideOutput) {
       send({
-        type: "LINE_RECEIVED",
-        line,
+        type: "OUTPUT_RECEIVED",
+        output,
       });
     }
   };
@@ -198,17 +196,14 @@ export const deployMachine = createMachine<
               APPROVE: {
                 target: "processing",
                 actions: send(
-                  { type: "SEND_INPUT", input: withNewline("yes") },
+                  { type: "SEND_LINE", input: "yes" },
                   { to: "pty" }
                 ),
               },
               REJECT: {
                 target: "processing",
                 actions: [
-                  send(
-                    { type: "SEND_INPUT", input: withNewline("no") },
-                    { to: "pty" }
-                  ),
+                  send({ type: "SEND_LINE", input: "no" }, { to: "pty" }),
                   assign<DeployContext, DeployEvent & { type: "REJECT" }>({
                     cancelled: true,
                   }),
@@ -235,17 +230,14 @@ export const deployMachine = createMachine<
               OVERRIDE: {
                 target: "processing",
                 actions: send(
-                  { type: "SEND_INPUT", input: withNewline("override") },
+                  { type: "SEND_LINE", input: "override" },
                   { to: "pty" }
                 ),
               },
               REJECT_OVERRIDE: {
                 target: "processing",
                 actions: [
-                  send(
-                    { type: "SEND_INPUT", input: withNewline("no") },
-                    { to: "pty" }
-                  ),
+                  send({ type: "SEND_LINE", input: "no" }, { to: "pty" }),
                   assign<
                     DeployContext,
                     DeployEvent & { type: "REJECT_OVERRIDE" }
@@ -265,63 +257,45 @@ export const deployMachine = createMachine<
   {
     services: {
       runTerraformInPty: (context, event) =>
-        terraformPtyService(context, event, spawnPtyFromEvent),
+        terraformPtyService(context, event, spawnPty),
     },
   }
 );
 
-function spawnPtyFromEvent(event: DeployEvent): pty.IPty {
-  if (event.type !== "START") {
-    throw Errors.Internal(
-      `Terraform CLI invocation state machine: Unexpected event caused transition to the running state: ${event.type}`
-    );
-  }
-
-  const { args, options } = event.pty;
-  const file =
-    os.platform() === "win32"
-      ? findExecutable(event.pty.file, options.cwd, options)
-      : event.pty.file;
-  logger.trace(
-    `Spawning pty with file=${file}, args=${
-      Array.isArray(args) ? `[${args.join(", ")}]` : `"${args}"`
-    }, options=${JSON.stringify(options)}`
-  );
-  const p = pty.spawn(file, args, options);
-
-  return p;
-}
-
 export function terraformPtyService(
   _context: DeployContext,
   event: DeployEvent,
-  spawnFunction: (event: DeployEvent) => pty.IPty
+  spawn = spawnPty
 ): (send: Sender<DeployEvent>, onReceive: Receiver<DeployEvent>) => void {
   return (send: Sender<DeployEvent>, onReceive: Receiver<DeployEvent>) => {
-    const p = spawnFunction(event);
+    if (event.type !== "START") {
+      throw Errors.Internal(
+        `Terraform CLI invocation state machine: Unexpected event caused transition to the running state: ${event.type}`
+      );
+    }
 
+    // Communication from the pty to the caller
+    const receiver = handleLineReceived(send);
+    const { exitCode, actions } = spawn(event.pty, (data) => {
+      receiver(data);
+    });
+
+    // Communication from the caller to the pty
     onReceive((event: DeployEvent) => {
-      if (event.type === "SEND_INPUT") {
-        p.write(event.input);
+      if (event.type === "SEND_LINE") {
+        actions.writeLine(event.input);
       }
     });
 
-    p.onData(handleLineReceived(send));
-
-    p.onExit(({ exitCode }) => {
+    exitCode.then((exitCode) => {
       send({ type: "EXITED", exitCode });
     });
 
     return () => {
       logger.trace("Terraform CLI state machine: cleaning up pty");
-      p.write("\x03"); // CTRL + C, pty.kill() does not work on windows
-      // TODO: is there a way to delay this? maybe go into a "killing" state first?
+      actions.stop();
     };
   };
-}
-
-function withNewline(str: string): string {
-  return `${str}${os.EOL}`;
 }
 
 export function createAndStartDeployService(options: {
@@ -427,66 +401,4 @@ export function createAndStartDestroyService(options: {
   service.send({ type: "START", pty: config });
 
   return service;
-}
-
-// src: https://github.com/Microsoft/vscode/blob/c0c9ea27d6e8d660d8716d7acee82cf3c00fa3e5/src/vs/workbench/parts/tasks/electron-browser/terminalTaskSystem.ts#L691
-// TODO: properly annotate source of this function
-function findExecutable(
-  command: string,
-  cwd: string,
-  options: { env?: { [key: string]: string } }
-): string {
-  // If we have an absolute path then we take it.
-  if (path.isAbsolute(command)) {
-    return command;
-  }
-  const dir = path.dirname(command);
-  if (dir !== ".") {
-    // We have a directory and the directory is relative (see above). Make the path absolute
-    // to the current working directory.
-    return path.join(cwd, command);
-  }
-  let paths: string[] | undefined = undefined;
-  // The options can override the PATH. So consider that PATH if present.
-  if (options && options.env) {
-    // Path can be named in many different ways and for the execution it doesn't matter
-    for (const key of Object.keys(options.env)) {
-      if (key.toLowerCase() === "path") {
-        if (typeof options.env[key] === "string") {
-          paths = options.env[key].split(path.delimiter);
-        }
-        break;
-      }
-    }
-  }
-  if (paths === void 0 && typeof process.env.PATH === "string") {
-    paths = process.env.PATH.split(path.delimiter);
-  }
-  // No PATH environment. Make path absolute to the cwd.
-  if (paths === void 0 || paths.length === 0) {
-    return path.join(cwd, command);
-  }
-  // We have a simple file name. We get the path variable from the env
-  // and try to find the executable on the path.
-  for (const pathEntry of paths) {
-    // The path entry is absolute.
-    let fullPath: string;
-    if (path.isAbsolute(pathEntry)) {
-      fullPath = path.join(pathEntry, command);
-    } else {
-      fullPath = path.join(cwd, pathEntry, command);
-    }
-    if (fs.existsSync(fullPath)) {
-      return fullPath;
-    }
-    let withExtension = fullPath + ".com";
-    if (fs.existsSync(withExtension)) {
-      return withExtension;
-    }
-    withExtension = fullPath + ".exe";
-    if (fs.existsSync(withExtension)) {
-      return withExtension;
-    }
-  }
-  return path.join(cwd, command);
 }

--- a/packages/@cdktf/cli-core/src/lib/models/pty-process.ts
+++ b/packages/@cdktf/cli-core/src/lib/models/pty-process.ts
@@ -18,8 +18,6 @@ export interface PtyProcessActions {
 }
 
 export interface PtyProcess {
-  // Resolves or rejects depending on the exit code of the process
-  progress: Promise<void>;
   // Always resolves, returns the exit code of the process
   exitCode: Promise<number>;
   actions: PtyProcessActions;
@@ -77,29 +75,20 @@ export function spawnPty(
 
   p.onData((data) => onData(data));
 
-  const progress = new Promise<void>((resolve, reject) => {
-    p.onExit(({ exitCode }) => {
-      if (exitCode === 0) {
-        resolve();
-      } else {
-        reject(
-          new Error(
-            `Pty (file=${file}, args=${
-              Array.isArray(args) ? `[${args.join(", ")}]` : `"${args}"`
-            }) exited with code ${exitCode}`
-          )
-        );
-      }
-    });
-  });
-
   const exitCode = new Promise<number>((resolve) => {
     p.onExit(({ exitCode }) => {
+      if (exitCode !== 0) {
+        logger.debug(
+          `Pty (file=${file}, args=${
+            Array.isArray(args) ? `[${args.join(", ")}]` : `"${args}"`
+          }) exited with code ${exitCode}`
+        );
+      }
       resolve(exitCode);
     });
   });
 
-  return { progress, exitCode, actions };
+  return { exitCode, actions };
 }
 
 // src: https://github.com/Microsoft/vscode/blob/c0c9ea27d6e8d660d8716d7acee82cf3c00fa3e5/src/vs/workbench/parts/tasks/electron-browser/terminalTaskSystem.ts#L691

--- a/packages/@cdktf/cli-core/src/lib/models/pty-process.ts
+++ b/packages/@cdktf/cli-core/src/lib/models/pty-process.ts
@@ -1,0 +1,165 @@
+import * as os from "os";
+import * as path from "path";
+import * as fs from "fs";
+
+import * as pty from "@cdktf/node-pty-prebuilt-multiarch";
+import { logger } from "@cdktf/commons";
+
+interface PtySpawnConfig {
+  file: Parameters<typeof pty.spawn>[0];
+  args: Parameters<typeof pty.spawn>[1];
+  options: Parameters<typeof pty.spawn>[2] & { cwd: string };
+}
+
+export interface PtyProcessActions {
+  write: (response: string) => void;
+  writeLine: (response: string) => void;
+  stop: () => void;
+}
+
+export interface PtyProcess {
+  // Resolves or rejects depending on the exit code of the process
+  progress: Promise<void>;
+  // Always resolves, returns the exit code of the process
+  exitCode: Promise<number>;
+  actions: PtyProcessActions;
+}
+
+/**
+ * A wrapper around node-pty that handles the platform specific differences
+ * and provide an intuitive API for bidirectional communication with the
+ * spawned process.
+ */
+export function spawnPty(
+  config: PtySpawnConfig,
+  onData: (data: string) => void
+): PtyProcess {
+  const { args, options } = config;
+  const file =
+    os.platform() === "win32"
+      ? findExecutable(config.file, options.cwd, options)
+      : config.file;
+
+  logger.trace(
+    `Spawning pty with file=${file}, args=${
+      Array.isArray(args) ? `[${args.join(", ")}]` : `"${args}"`
+    }, options=${JSON.stringify(options)}`
+  );
+
+  const p = pty.spawn(file, args, options);
+
+  const actions: PtyProcessActions = {
+    write: (response: string) => {
+      logger.trace(
+        `Sending response to pty (file=${file}, args=${
+          Array.isArray(args) ? `[${args.join(", ")}]` : `"${args}"`
+        }): ${response}`
+      );
+      p.write(response);
+    },
+    writeLine: (response: string) => {
+      logger.trace(
+        `Sending response (with newline) to pty (file=${file}, args=${
+          Array.isArray(args) ? `[${args.join(", ")}]` : `"${args}"`
+        }): ${response}`
+      );
+      p.write(`${response}${os.EOL}`);
+    },
+    stop: () => {
+      logger.trace(
+        `Aborting pty (file=${file}, args=${
+          Array.isArray(args) ? `[${args.join(", ")}]` : `"${args}"`
+        })`
+      );
+      p.write("\x03"); // CTRL + C, pty.kill() does not work on windows
+    },
+  };
+
+  p.onData((data) => onData(data));
+
+  const progress = new Promise<void>((resolve, reject) => {
+    p.onExit(({ exitCode }) => {
+      if (exitCode === 0) {
+        resolve();
+      } else {
+        reject(
+          new Error(
+            `Pty (file=${file}, args=${
+              Array.isArray(args) ? `[${args.join(", ")}]` : `"${args}"`
+            }) exited with code ${exitCode}`
+          )
+        );
+      }
+    });
+  });
+
+  const exitCode = new Promise<number>((resolve) => {
+    p.onExit(({ exitCode }) => {
+      resolve(exitCode);
+    });
+  });
+
+  return { progress, exitCode, actions };
+}
+
+// src: https://github.com/Microsoft/vscode/blob/c0c9ea27d6e8d660d8716d7acee82cf3c00fa3e5/src/vs/workbench/parts/tasks/electron-browser/terminalTaskSystem.ts#L691
+// TODO: properly annotate source of this function
+function findExecutable(
+  command: string,
+  cwd: string,
+  options: { env?: { [key: string]: string } }
+): string {
+  // If we have an absolute path then we take it.
+  if (path.isAbsolute(command)) {
+    return command;
+  }
+  const dir = path.dirname(command);
+  if (dir !== ".") {
+    // We have a directory and the directory is relative (see above). Make the path absolute
+    // to the current working directory.
+    return path.join(cwd, command);
+  }
+  let paths: string[] | undefined = undefined;
+  // The options can override the PATH. So consider that PATH if present.
+  if (options && options.env) {
+    // Path can be named in many different ways and for the execution it doesn't matter
+    for (const key of Object.keys(options.env)) {
+      if (key.toLowerCase() === "path") {
+        if (typeof options.env[key] === "string") {
+          paths = options.env[key].split(path.delimiter);
+        }
+        break;
+      }
+    }
+  }
+  if (paths === void 0 && typeof process.env.PATH === "string") {
+    paths = process.env.PATH.split(path.delimiter);
+  }
+  // No PATH environment. Make path absolute to the cwd.
+  if (paths === void 0 || paths.length === 0) {
+    return path.join(cwd, command);
+  }
+  // We have a simple file name. We get the path variable from the env
+  // and try to find the executable on the path.
+  for (const pathEntry of paths) {
+    // The path entry is absolute.
+    let fullPath: string;
+    if (path.isAbsolute(pathEntry)) {
+      fullPath = path.join(pathEntry, command);
+    } else {
+      fullPath = path.join(cwd, pathEntry, command);
+    }
+    if (fs.existsSync(fullPath)) {
+      return fullPath;
+    }
+    let withExtension = fullPath + ".com";
+    if (fs.existsSync(withExtension)) {
+      return withExtension;
+    }
+    withExtension = fullPath + ".exe";
+    if (fs.existsSync(withExtension)) {
+      return withExtension;
+    }
+  }
+  return path.join(cwd, command);
+}

--- a/packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts
+++ b/packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts
@@ -122,8 +122,13 @@ export class TerraformCli implements Terraform {
       args.push("-no-color");
     }
 
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    let initCanNotContinue = (_err: any) => {};
+    const rejectsIfInitCanNotContinue = new Promise((_resolve, reject) => {
+      initCanNotContinue = reject;
+    });
+
     const stdout = this.onStdout("init");
-    const stderr = this.onStderr("init");
     const { actions, progress } = spawnPty(
       {
         file: terraformBinaryName,
@@ -139,8 +144,8 @@ export class TerraformCli implements Terraform {
           if (migrateState) {
             actions.writeLine("yes");
           } else {
-            actions.writeLine("no");
-            stderr(
+            actions.stop();
+            initCanNotContinue(
               "Please pass the --migrate-state flag to migrate your state"
             );
           }
@@ -151,7 +156,7 @@ export class TerraformCli implements Terraform {
       actions.stop();
     });
 
-    await progress;
+    await Promise.race([progress, rejectsIfInitCanNotContinue]);
 
     // TODO: this might have performance implications because we don't know if we're
     // running a remote plan or a local one (so we run it always for all platforms)

--- a/packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts
+++ b/packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts
@@ -129,7 +129,7 @@ export class TerraformCli implements Terraform {
     });
 
     const stdout = this.onStdout("init");
-    const { actions, progress } = spawnPty(
+    const { actions, exitCode } = spawnPty(
       {
         file: terraformBinaryName,
         args,
@@ -156,6 +156,11 @@ export class TerraformCli implements Terraform {
       actions.stop();
     });
 
+    const progress = exitCode.then((code) => {
+      if (code !== 0) {
+        throw new Error(`terraform init failed with exit code ${code}`);
+      }
+    });
     await Promise.race([progress, rejectsIfInitCanNotContinue]);
 
     // TODO: this might have performance implications because we don't know if we're

--- a/packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts
+++ b/packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts
@@ -25,6 +25,7 @@ import { waitFor } from "xstate/lib/waitFor";
 import { missingVariable } from "../errors";
 import { terraformJsonSchema } from "../terraform-json";
 import { AbortSignal } from "node-abort-controller"; // polyfill until we update to node 16
+import { spawnPty } from "./pty-process";
 
 export class TerraformCliPlan
   extends AbstractTerraformPlan
@@ -106,10 +107,14 @@ export class TerraformCli implements Terraform {
         createTerraformLogHandler(phase, filter)(stderr.toString(), true);
   }
 
-  public async init(needsUpgrade: boolean, noColor?: boolean): Promise<void> {
+  public async init(
+    needsUpgrade: boolean,
+    noColor: boolean,
+    migrateState: boolean
+  ): Promise<void> {
     await this.setUserAgent();
 
-    const args = ["init", "-input=false"];
+    const args = ["init"];
     if (needsUpgrade) {
       args.push("-upgrade");
     }
@@ -117,17 +122,36 @@ export class TerraformCli implements Terraform {
       args.push("-no-color");
     }
 
-    await exec(
-      terraformBinaryName,
-      args,
+    const stdout = this.onStdout("init");
+    const stderr = this.onStderr("init");
+    const { actions, progress } = spawnPty(
       {
-        cwd: this.workdir,
-        env: process.env,
-        signal: this.abortSignal,
+        file: terraformBinaryName,
+        args,
+        options: {
+          cwd: this.workdir,
+          env: process.env as any,
+        },
       },
-      this.onStdout("init"),
-      this.onStderr("init")
+      (data) => {
+        stdout(data);
+        if (data.includes("Should Terraform migrate your existing state?")) {
+          if (migrateState) {
+            actions.writeLine("yes");
+          } else {
+            actions.writeLine("no");
+            stderr(
+              "Please pass the --migrate-state flag to migrate your state"
+            );
+          }
+        }
+      }
     );
+    this.abortSignal.addEventListener("abort", () => {
+      actions.stop();
+    });
+
+    await progress;
 
     // TODO: this might have performance implications because we don't know if we're
     // running a remote plan or a local one (so we run it always for all platforms)

--- a/packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts
+++ b/packages/@cdktf/cli-core/src/lib/models/terraform-cli.ts
@@ -291,8 +291,8 @@ export class TerraformCli implements Terraform {
       logger.trace(
         `Terraform CLI state machine event: ${JSON.stringify(event)}`
       );
-      if (isDeployEvent(event, "LINE_RECEIVED"))
-        this.onStdout(type)(event.line);
+      if (isDeployEvent(event, "OUTPUT_RECEIVED"))
+        this.onStdout(type)(event.output);
       else if (isDeployEvent(event, "APPROVED_EXTERNALLY"))
         callback({ type: "external approval reply", approved: true });
       else if (isDeployEvent(event, "REJECTED_EXTERNALLY"))

--- a/packages/@cdktf/cli-core/src/lib/models/terraform.ts
+++ b/packages/@cdktf/cli-core/src/lib/models/terraform.ts
@@ -129,7 +129,11 @@ export type TerraformDeployState =
   | { type: "external sentinel override reply"; overridden: boolean };
 
 export interface Terraform {
-  init: (upgrade: boolean, noColor?: boolean) => Promise<void>;
+  init: (
+    upgrade: boolean,
+    noColor: boolean,
+    migrateState: boolean
+  ) => Promise<void>;
   plan: (opts: {
     destroy: boolean;
     refreshOnly?: boolean;

--- a/packages/@cdktf/cli-core/src/test/lib/cdktf-project.test.ts
+++ b/packages/@cdktf/cli-core/src/test/lib/cdktf-project.test.ts
@@ -403,11 +403,17 @@ describe("CdktfProject", () => {
       });
 
       try {
-        await cdktfProject.deploy({
-          stackNames: ["stack1", "stack2", "stack3"],
-          parallelism: 100,
-          autoApprove: true,
-        });
+        await cdktfProject
+          .deploy({
+            stackNames: ["stack1", "stack2", "stack3"],
+            parallelism: 100,
+            autoApprove: true,
+          })
+          .catch((e) => {
+            console.log("deploy throws", e);
+            throw e;
+          });
+
         throw new Error("This error should not be thrown");
       } catch (e) {
         expect(e).toMatchInlineSnapshot(

--- a/packages/@cdktf/cli-core/src/test/models/deploy-machine.test.ts
+++ b/packages/@cdktf/cli-core/src/test/models/deploy-machine.test.ts
@@ -1,7 +1,6 @@
 import { interpret } from "xstate";
-import { IPty } from "@cdktf/node-pty-prebuilt-multiarch";
+import { spawnPty } from "../../lib/models/pty-process";
 import {
-  DeployEvent,
   deployMachine,
   extractVariableNameFromPrompt,
   terraformPtyService,
@@ -84,30 +83,41 @@ describe("transitions", () => {
   });
 });
 
+function mockPty(ptyEvents: string[]): typeof spawnPty {
+  const actions = {
+    write: jest.fn(),
+    writeLine: jest.fn(),
+    stop: jest.fn(),
+  };
+  return (_config: any, onData: (data: string) => void) => {
+    ptyEvents.forEach((event, index) => {
+      setTimeout(() => {
+        onData(event);
+      }, 100 * (index + 1));
+    });
+
+    return {
+      actions,
+      progress: new Promise((resolve) => {
+        setTimeout(() => resolve(), ptyEvents.length * 200);
+      }),
+      exitCode: new Promise((resolve) => {
+        setTimeout(() => resolve(0), ptyEvents.length * 200);
+      }),
+    };
+  };
+}
+
 describe("pty events", () => {
   it("transitions when pty receives an approval question", (done) => {
-    function mockPty(_event: DeployEvent) {
-      const onDataListeners: any[] = [];
-      const onExitListeners: any[] = [];
-      const pty = {
-        onData: (listener: (e: string) => any) =>
-          onDataListeners.push(listener),
-        onExit: (listener: (e: string) => any) =>
-          onExitListeners.push(listener),
-      } as unknown as IPty;
-
-      setTimeout(() => {
-        onDataListeners.forEach((listener) =>
-          listener(`Do you want to perform these actions\nEnter a value:`)
-        );
-      }, 100);
-
-      return pty;
-    }
     const mockDeployMachine = deployMachine.withConfig({
       services: {
         runTerraformInPty: (context, event) =>
-          terraformPtyService(context, event, mockPty),
+          terraformPtyService(
+            context,
+            event,
+            mockPty([`Do you want to perform these actions\nEnter a value:`])
+          ),
       },
     });
 
@@ -132,31 +142,19 @@ describe("pty events", () => {
   });
 
   it("transitions when pty receives an override question", (done) => {
-    function mockPty(_event: DeployEvent) {
-      const onDataListeners: any[] = [];
-      const onExitListeners: any[] = [];
-      const pty = {
-        onData: (listener: (e: string) => any) =>
-          onDataListeners.push(listener),
-        onExit: (listener: (e: string) => any) =>
-          onExitListeners.push(listener),
-      } as unknown as IPty;
-
-      setTimeout(() => {
-        onDataListeners.forEach((listener) =>
-          listener(`Do you want to override the soft failed policy check?
-          Only 'override' will be accepted to override.
-          
-          Enter a value:`)
-        );
-      }, 100);
-
-      return pty;
-    }
     const mockDeployMachine = deployMachine.withConfig({
       services: {
         runTerraformInPty: (context, event) =>
-          terraformPtyService(context, event, mockPty),
+          terraformPtyService(
+            context,
+            event,
+            mockPty([
+              `Do you want to override the soft failed policy check?
+          Only 'override' will be accepted to override.
+          
+          Enter a value:`,
+            ])
+          ),
       },
     });
 
@@ -181,38 +179,20 @@ describe("pty events", () => {
   });
 
   it("transitions to rejected state when done externally", (done) => {
-    function mockPty(_event: DeployEvent) {
-      const onDataListeners: any[] = [];
-      const onExitListeners: any[] = [];
-      const pty = {
-        onData: (listener: (e: string) => any) =>
-          onDataListeners.push(listener),
-        onExit: (listener: (e: string) => any) =>
-          onExitListeners.push(listener),
-        write: jest.fn(),
-      } as unknown as IPty;
-
-      setTimeout(() => {
-        onDataListeners.forEach((listener) =>
-          listener(`Do you want to override the soft failed policy check?
-          Only 'override' will be accepted to override.
-          
-          Enter a value:`)
-        );
-      }, 100);
-
-      setTimeout(() => {
-        onDataListeners.forEach((listener) =>
-          listener(`discarded using the UI or API`)
-        );
-      }, 200);
-
-      return pty;
-    }
     const mockDeployMachine = deployMachine.withConfig({
       services: {
         runTerraformInPty: (context, event) =>
-          terraformPtyService(context, event, mockPty),
+          terraformPtyService(
+            context,
+            event,
+            mockPty([
+              `Do you want to override the soft failed policy check?
+          Only 'override' will be accepted to override.
+          
+          Enter a value:`,
+              `discarded using the UI or API`,
+            ])
+          ),
       },
     });
 
@@ -252,8 +232,8 @@ describe("handleLineReceived", () => {
     const send = jest.fn();
     handleLineReceived(send)("foo");
     expect(send).toHaveBeenCalledWith({
-      type: "LINE_RECEIVED",
-      line: "foo",
+      type: "OUTPUT_RECEIVED",
+      Line: "foo",
     });
   });
 
@@ -265,8 +245,8 @@ describe("handleLineReceived", () => {
     `;
     handleLineReceived(send)(input);
     expect(send).not.toHaveBeenCalledWith({
-      type: "LINE_RECEIVED",
-      line: input,
+      type: "OUTPUT_RECEIVED",
+      Line: input,
     });
     expect(send).toHaveBeenCalledWith({
       type: "VARIABLE_MISSING",

--- a/packages/@cdktf/cli-core/src/test/models/deploy-machine.test.ts
+++ b/packages/@cdktf/cli-core/src/test/models/deploy-machine.test.ts
@@ -233,7 +233,7 @@ describe("handleLineReceived", () => {
     handleLineReceived(send)("foo");
     expect(send).toHaveBeenCalledWith({
       type: "OUTPUT_RECEIVED",
-      Line: "foo",
+      output: "foo",
     });
   });
 
@@ -246,7 +246,7 @@ describe("handleLineReceived", () => {
     handleLineReceived(send)(input);
     expect(send).not.toHaveBeenCalledWith({
       type: "OUTPUT_RECEIVED",
-      Line: input,
+      output: input,
     });
     expect(send).toHaveBeenCalledWith({
       type: "VARIABLE_MISSING",

--- a/packages/cdktf-cli/src/bin/cmds/deploy.ts
+++ b/packages/cdktf-cli/src/bin/cmds/deploy.ts
@@ -84,6 +84,12 @@ class Command extends BaseCommand {
         required: false,
         desc: "Disables terminal formatting sequences in the output.",
       })
+      .option("migrate-state", {
+        type: "boolean",
+        default: false,
+        required: false,
+        desc: "Reconfigure a backend, and attempt to migrate any existing state for all selected stacks.",
+      })
       .option("var", {
         type: "array",
         default: [],

--- a/packages/cdktf-cli/src/bin/cmds/deploy.ts
+++ b/packages/cdktf-cli/src/bin/cmds/deploy.ts
@@ -88,7 +88,7 @@ class Command extends BaseCommand {
         type: "boolean",
         default: false,
         required: false,
-        desc: "Reconfigure a backend, and attempt to migrate any existing state for all selected stacks.",
+        desc: "Pass this flag after switching state backends to approve a state migration for all targeted stacks",
       })
       .option("var", {
         type: "array",

--- a/packages/cdktf-cli/src/bin/cmds/destroy.ts
+++ b/packages/cdktf-cli/src/bin/cmds/destroy.ts
@@ -77,7 +77,7 @@ class Command extends BaseCommand {
         type: "boolean",
         default: false,
         required: false,
-        desc: "Reconfigure a backend, and attempt to migrate any existing state for all selected stacks.",
+        desc: "Pass this flag after switching state backends to approve a state migration for all targeted stacks",
       })
       .showHelpOnFail(true);
 

--- a/packages/cdktf-cli/src/bin/cmds/destroy.ts
+++ b/packages/cdktf-cli/src/bin/cmds/destroy.ts
@@ -73,6 +73,12 @@ class Command extends BaseCommand {
         required: false,
         desc: "Disables terminal formatting sequences in the output.",
       })
+      .option("migrate-state", {
+        type: "boolean",
+        default: false,
+        required: false,
+        desc: "Reconfigure a backend, and attempt to migrate any existing state for all selected stacks.",
+      })
       .showHelpOnFail(true);
 
   public async handleCommand(argv: any) {

--- a/packages/cdktf-cli/src/bin/cmds/diff.ts
+++ b/packages/cdktf-cli/src/bin/cmds/diff.ts
@@ -69,7 +69,7 @@ class Command extends BaseCommand {
         type: "boolean",
         default: false,
         required: false,
-        desc: "Reconfigure a backend, and attempt to migrate any existing state for the selected stack.",
+        desc: "Pass this flag after switching state backends to approve a state migration for the targeted stack",
       })
       .showHelpOnFail(true);
 

--- a/packages/cdktf-cli/src/bin/cmds/diff.ts
+++ b/packages/cdktf-cli/src/bin/cmds/diff.ts
@@ -65,6 +65,12 @@ class Command extends BaseCommand {
         required: false,
         desc: "Disables terminal formatting sequences in the output.",
       })
+      .option("migrate-state", {
+        type: "boolean",
+        default: false,
+        required: false,
+        desc: "Reconfigure a backend, and attempt to migrate any existing state for the selected stack.",
+      })
       .showHelpOnFail(true);
 
   public async handleCommand(argv: any) {

--- a/packages/cdktf-cli/src/bin/cmds/handlers.ts
+++ b/packages/cdktf-cli/src/bin/cmds/handlers.ts
@@ -138,6 +138,7 @@ export async function deploy(argv: any) {
   const vars = argv.var;
   const varFiles = sanitizeVarFiles(argv.varFile);
   const noColor = argv.noColor;
+  const migrateState = argv.migrateState;
 
   let outputsPath: string | undefined = undefined;
   // eslint-disable-next-line @typescript-eslint/no-empty-function
@@ -163,6 +164,7 @@ export async function deploy(argv: any) {
       vars,
       varFiles,
       noColor,
+      migrateState,
     })
   );
 }
@@ -183,6 +185,7 @@ export async function destroy(argv: any) {
   const vars = argv.var;
   const varFiles = sanitizeVarFiles(argv.varFile);
   const noColor = argv.noColor;
+  const migrateState = argv.migrateState;
 
   await renderInk(
     React.createElement(Destroy, {
@@ -196,6 +199,7 @@ export async function destroy(argv: any) {
       vars,
       varFiles,
       noColor,
+      migrateState,
     })
   );
 }
@@ -213,6 +217,7 @@ export async function diff(argv: any) {
   const vars = argv.var;
   const varFiles = sanitizeVarFiles(argv.varFile);
   const noColor = argv.noColor;
+  const migrateState = argv.migrateState;
 
   await renderInk(
     React.createElement(Diff, {
@@ -224,6 +229,7 @@ export async function diff(argv: any) {
       vars,
       varFiles,
       noColor,
+      migrateState,
     })
   );
 }

--- a/packages/cdktf-cli/src/bin/cmds/ui/deploy.tsx
+++ b/packages/cdktf-cli/src/bin/cmds/ui/deploy.tsx
@@ -60,6 +60,7 @@ interface DeployConfig {
   vars?: string[];
   varFiles?: string[];
   noColor?: boolean;
+  migrateState?: boolean;
 }
 
 export const Deploy = ({
@@ -76,6 +77,7 @@ export const Deploy = ({
   vars,
   varFiles,
   noColor,
+  migrateState,
 }: DeployConfig): React.ReactElement => {
   const [outputs, setOutputs] = useState<NestedTerraformOutputs>();
   const { status, logEntries } = useCdktfProject(
@@ -91,6 +93,7 @@ export const Deploy = ({
         vars,
         varFiles,
         noColor,
+        migrateState,
       });
 
       if (onOutputsRetrieved) {

--- a/packages/cdktf-cli/src/bin/cmds/ui/destroy.tsx
+++ b/packages/cdktf-cli/src/bin/cmds/ui/destroy.tsx
@@ -18,6 +18,7 @@ interface DestroyConfig {
   parallelism?: number;
   terraformParallelism?: number;
   noColor?: boolean;
+  migrateState?: boolean;
   vars?: string[];
   varFiles?: string[];
 }
@@ -31,6 +32,7 @@ export const Destroy = ({
   parallelism,
   terraformParallelism,
   noColor,
+  migrateState,
   vars,
   varFiles,
 }: DestroyConfig): React.ReactElement => {
@@ -44,6 +46,7 @@ export const Destroy = ({
         parallelism,
         terraformParallelism,
         noColor,
+        migrateState,
         vars,
         varFiles,
       })

--- a/packages/cdktf-cli/src/bin/cmds/ui/diff.tsx
+++ b/packages/cdktf-cli/src/bin/cmds/ui/diff.tsx
@@ -12,6 +12,7 @@ interface DiffConfig {
   vars?: string[];
   varFiles?: string[];
   noColor?: boolean;
+  migrateState?: boolean;
 }
 
 export const Diff = ({
@@ -23,6 +24,7 @@ export const Diff = ({
   vars,
   varFiles,
   noColor,
+  migrateState,
 }: DiffConfig): React.ReactElement => {
   const { status, logEntries } = useCdktfProject(
     { outDir, synthCommand },
@@ -34,6 +36,7 @@ export const Diff = ({
         vars,
         varFiles,
         noColor,
+        migrateState,
       })
   );
 

--- a/test/typescript/terraform-cloud/test.ts
+++ b/test/typescript/terraform-cloud/test.ts
@@ -77,9 +77,15 @@ describe("full integration test", () => {
     });
 
     process.env.TF_EXECUTE_LOCAL = "true";
-    await driver.deploy(["source-stack"]);
+    await driver.deploy(["source-stack"], "before-migration.json");
     process.env.TF_EXECUTE_LOCAL = undefined;
-    await driver.deploy(["source-stack"]);
+    await driver.deploy(["source-stack"], "after-migration.json", [
+      "--migrate-state",
+    ]);
+
+    expect(readFileSync("before-migration.json")).toEqual(
+      readFileSync("after-migration.json")
+    );
 
     await client.Workspaces.deleteByName(orgName, workspaceName);
   });

--- a/website/docs/cdktf/cli-reference/commands.mdx
+++ b/website/docs/cdktf/cli-reference/commands.mdx
@@ -147,7 +147,7 @@ Options:
       --version                                 Show version number  [boolean]
       --disable-plugin-cache-env                Dont set TF_PLUGIN_CACHE_DIR automatically. This is useful when the plugin cache is configured differently. Supported using the env CDKTF_DISABLE_PLUGIN_CACHE_ENV.  [boolean] [default: false]
       --log-level                               Which log level should be written. Only supported via setting the env CDKTF_LOG_LEVEL  [string]
-  -a, --app                                     Command to use in order to execute cdktf app  [required]
+  -a, --app                                     Command to use in order to execute cdktf app  [required] [default: "npx ts-node main.ts"]
   -o, --output                                  Output directory for the synthesized Terraform config  [required] [default: "cdktf.out"]
       --auto-approve                            Auto approve  [boolean] [default: false]
       --outputs-file                            Path to file where stack outputs will be written as JSON  [string]
@@ -156,6 +156,10 @@ Options:
       --parallelism                             Number of concurrent CDKTF stacks to run. Defaults to infinity, denoted by -1  [number] [default: -1]
       --refresh-only                            Select the "refresh only" planning mode, which checks whether remote objects still match the outcome of the most recent Terraform apply but does not propose any actions to undo any changes made outside of Terraform.  [boolean] [default: false]
       --terraform-parallelism                   Forwards value as the `-parallelism` flag to Terraform. By default, the this flag is not forwarded to Terraform. Note: This flag is not supported by remote / cloud backend  [number] [default: -1]
+      --no-color                                Disables terminal formatting sequences in the output.  [boolean] [default: false]
+      --migrate-state                           Reconfigure a backend, and attempt to migrate any existing state for all selected stacks.  [boolean] [default: false]
+      --var                                     Set a value for one of the input variables in the stack or stacks to apply. Use this option more than once to set more than one variable.  [array] [default: []]
+      --var-file                                Load variable values from the given file, in addition to the default files terraform.tfvars and *.auto.tfvars. Use this option more than once to include more than one variables file.  [array] [default: []]
   -h, --help                                    Show help  [boolean]
 ```
 
@@ -213,24 +217,23 @@ cdktf destroy [stacks..]
 Destroy the given stacks
 
 Positionals:
-  stacks  Destroy stacks matching the given ids. Required when more than one stack is present in the app                                                                                     [array] [default: []]
+  stacks  Destroy stacks matching the given ids. Required when more than one stack is present in the app  [array] [default: []]
 
 Options:
-      --version                            Show version number                                                                                                                                           [boolean]
-      --disable-plugin-cache-env           Dont set TF_PLUGIN_CACHE_DIR automatically. This is useful when the plugin cache is configured differently. Supported using the env CDKTF_DISABLE_PLUGIN_CACHE_ENV.
-                                                                                                                                                                                        [boolean] [default: false]
-      --log-level                          Which log level should be written. Only supported via setting the env CDKTF_LOG_LEVEL                                                                          [string]
-  -a, --app                                Command to use in order to execute cdktf app                                                                                                                 [required]
-  -o, --output                             Output directory for the synthesized Terraform config                                                                                 [required] [default: "cdktf.out"]
-      --auto-approve                       Auto approve                                                                                                                                 [boolean] [default: false]
-      --ignore-missing-stack-dependencies  Don't check if all stacks specified in the command have their dependencies included as well                                                  [boolean] [default: false]
-      --parallelism                        Number of concurrent CDKTF stacks to run. Defaults to infinity, denoted by -1                                                                    [number] [default: -1]
-      --terraform-parallelism              Forwards value as the `-parallelism` flag to Terraform. By default, the this flag is not forwarded to Terraform. Note: This flag is not supported by remote / cloud
-                                           backend                                                                                                                                          [number] [default: -1]
-      --var                                Set a value for one of the input variables in the stack or stacks to apply. Use this option more than once to set more than one variable.         [array] [default: []]
-      --var-file                           Load variable values from the given file, in addition to the default files terraform.tfvars and *.auto.tfvars. Use this option more than once to include more than one
-                                           variables file.                                                                                                                                   [array] [default: []]
-  -h, --help                               Show help                                                                                                                                                     [boolean]
+      --version                            Show version number  [boolean]
+      --disable-plugin-cache-env           Dont set TF_PLUGIN_CACHE_DIR automatically. This is useful when the plugin cache is configured differently. Supported using the env CDKTF_DISABLE_PLUGIN_CACHE_ENV.  [boolean] [default: false]
+      --log-level                          Which log level should be written. Only supported via setting the env CDKTF_LOG_LEVEL  [string]
+  -a, --app                                Command to use in order to execute cdktf app  [required] [default: "npx ts-node main.ts"]
+  -o, --output                             Output directory for the synthesized Terraform config  [required] [default: "cdktf.out"]
+      --auto-approve                       Auto approve  [boolean] [default: false]
+      --ignore-missing-stack-dependencies  Don't check if all stacks specified in the command have their dependencies included as well  [boolean] [default: false]
+      --parallelism                        Number of concurrent CDKTF stacks to run. Defaults to infinity, denoted by -1  [number] [default: -1]
+      --terraform-parallelism              Forwards value as the `-parallelism` flag to Terraform. By default, the this flag is not forwarded to Terraform. Note: This flag is not supported by remote / cloud backend  [number] [default: -1]
+      --var                                Set a value for one of the input variables in the stack or stacks to apply. Use this option more than once to set more than one variable.  [array] [default: []]
+      --var-file                           Load variable values from the given file, in addition to the default files terraform.tfvars and *.auto.tfvars. Use this option more than once to include more than one variables file.  [array] [default: []]
+      --no-color                           Disables terminal formatting sequences in the output.  [boolean] [default: false]
+      --migrate-state                      Reconfigure a backend, and attempt to migrate any existing state for all selected stacks.  [boolean] [default: false]
+  -h, --help                               Show help  [boolean]
 ```
 
 ~> **Note:** The `parallelism` flag has a different behavior than the [terraform parallelism flag](/cli/commands/apply#parallelism-n). To set the custom terraform parallelism flag, please use the `--terraform-parallelism` flag instead.
@@ -285,23 +288,21 @@ cdktf diff [stack]
 Perform a diff (terraform plan) for the given stack
 
 Positionals:
-  stack  Diff stack which matches the given id only. Required when more than one stack is present in the app                                                                                              [string]
+  stack  Diff stack which matches the given id only. Required when more than one stack is present in the app  [string]
 
 Options:
-      --version                   Show version number                                                                                                                                                    [boolean]
-      --disable-plugin-cache-env  Dont set TF_PLUGIN_CACHE_DIR automatically. This is useful when the plugin cache is configured differently. Supported using the env CDKTF_DISABLE_PLUGIN_CACHE_ENV.
-                                                                                                                                                                                        [boolean] [default: false]
-      --log-level                 Which log level should be written. Only supported via setting the env CDKTF_LOG_LEVEL                                                                                   [string]
-  -a, --app                       Command to use in order to execute cdktf app                                                                                                                          [required]
-  -o, --output                    Output directory for the synthesized Terraform config                                                                                          [required] [default: "cdktf.out"]
-      --refresh-only              Select the "refresh only" planning mode, which checks whether remote objects still match the outcome of the most recent Terraform apply but does not propose any actions to undo
-                                  any changes made outside of Terraform.                                                                                                                [boolean] [default: false]
-      --terraform-parallelism     Forwards value as the `-parallelism` flag to Terraform. By default, the this flag is not forwarded to Terraform. Note: This flag is not supported by remote / cloud backend
-                                                                                                                                                                                            [number] [default: -1]
-      --var                       Set a value for one of the input variables in the stack. Use this option more than once to set more than one variable.                                     [array] [default: []]
-      --var-file                  Load variable values from the given file, in addition to the default files terraform.tfvars and *.auto.tfvars. Use this option more than once to include more than one variables
-                                  file.                                                                                                                                                      [array] [default: []]
-  -h, --help                      Show help                                                                                                                                                              [boolean]
+      --version                   Show version number  [boolean]
+      --disable-plugin-cache-env  Dont set TF_PLUGIN_CACHE_DIR automatically. This is useful when the plugin cache is configured differently. Supported using the env CDKTF_DISABLE_PLUGIN_CACHE_ENV.  [boolean] [default: false]
+      --log-level                 Which log level should be written. Only supported via setting the env CDKTF_LOG_LEVEL  [string]
+  -a, --app                       Command to use in order to execute cdktf app  [required] [default: "npx ts-node main.ts"]
+  -o, --output                    Output directory for the synthesized Terraform config  [required] [default: "cdktf.out"]
+      --refresh-only              Select the "refresh only" planning mode, which checks whether remote objects still match the outcome of the most recent Terraform apply but does not propose any actions to undo any changes made outside of Terraform.  [boolean] [default: false]
+      --terraform-parallelism     Forwards value as the `-parallelism` flag to Terraform. By default, the this flag is not forwarded to Terraform. Note: This flag is not supported by remote / cloud backend  [number] [default: -1]
+      --var                       Set a value for one of the input variables in the stack. Use this option more than once to set more than one variable.  [array] [default: []]
+      --var-file                  Load variable values from the given file, in addition to the default files terraform.tfvars and *.auto.tfvars. Use this option more than once to include more than one variables file.  [array] [default: []]
+      --no-color                  Disables terminal formatting sequences in the output.  [boolean] [default: false]
+      --migrate-state             Reconfigure a backend, and attempt to migrate any existing state for the selected stack.  [boolean] [default: false]
+  -h, --help                      Show help  [boolean]
 ```
 
 ~> **Note:** The `parallelism` flag has a different behavior than the [terraform parallelism flag](/cli/commands/plan#parallelism-n). To set the custom terraform parallelism flag, please use the `--terraform-parallelism` flag instead.

--- a/website/docs/cdktf/cli-reference/commands.mdx
+++ b/website/docs/cdktf/cli-reference/commands.mdx
@@ -157,7 +157,7 @@ Options:
       --refresh-only                            Select the "refresh only" planning mode, which checks whether remote objects still match the outcome of the most recent Terraform apply but does not propose any actions to undo any changes made outside of Terraform.  [boolean] [default: false]
       --terraform-parallelism                   Forwards value as the `-parallelism` flag to Terraform. By default, the this flag is not forwarded to Terraform. Note: This flag is not supported by remote / cloud backend  [number] [default: -1]
       --no-color                                Disables terminal formatting sequences in the output.  [boolean] [default: false]
-      --migrate-state                           Reconfigure a backend, and attempt to migrate any existing state for all selected stacks.  [boolean] [default: false]
+      --migrate-state                           Pass this flag after switching state backends to approve a state migration for all targeted stacks  [boolean] [default: false]
       --var                                     Set a value for one of the input variables in the stack or stacks to apply. Use this option more than once to set more than one variable.  [array] [default: []]
       --var-file                                Load variable values from the given file, in addition to the default files terraform.tfvars and *.auto.tfvars. Use this option more than once to include more than one variables file.  [array] [default: []]
   -h, --help                                    Show help  [boolean]
@@ -232,7 +232,7 @@ Options:
       --var                                Set a value for one of the input variables in the stack or stacks to apply. Use this option more than once to set more than one variable.  [array] [default: []]
       --var-file                           Load variable values from the given file, in addition to the default files terraform.tfvars and *.auto.tfvars. Use this option more than once to include more than one variables file.  [array] [default: []]
       --no-color                           Disables terminal formatting sequences in the output.  [boolean] [default: false]
-      --migrate-state                      Reconfigure a backend, and attempt to migrate any existing state for all selected stacks.  [boolean] [default: false]
+      --migrate-state                      Pass this flag after switching state backends to approve a state migration for all targeted stacks  [boolean] [default: false]
   -h, --help                               Show help  [boolean]
 ```
 
@@ -301,7 +301,7 @@ Options:
       --var                       Set a value for one of the input variables in the stack. Use this option more than once to set more than one variable.  [array] [default: []]
       --var-file                  Load variable values from the given file, in addition to the default files terraform.tfvars and *.auto.tfvars. Use this option more than once to include more than one variables file.  [array] [default: []]
       --no-color                  Disables terminal formatting sequences in the output.  [boolean] [default: false]
-      --migrate-state             Reconfigure a backend, and attempt to migrate any existing state for the selected stack.  [boolean] [default: false]
+      --migrate-state             Pass this flag after switching state backends to approve a state migration for the targeted stack  [boolean] [default: false]
   -h, --help                      Show help  [boolean]
 ```
 

--- a/website/docs/cdktf/concepts/remote-backends.mdx
+++ b/website/docs/cdktf/concepts/remote-backends.mdx
@@ -345,53 +345,30 @@ func main() {
 
 </CodeTabs>
 
-3. Run `cdktf synth` to generate the Terraform configuration file.
-
-4. Move the Terraform state file into the output directory.
+3. Run `cdktf diff <stack name> --migrate-state` to migrate the state into Terraform Cloud / Terraform Enterprise.
 
    ```bash
-   mv terraform.hello-terraform.tfstate cdktf.out/stacks/hello-terraform
+   Initializing Terraform Cloud...
+   Migrating from backend "local" to Terraform Cloud.
+   Do you wish to proceed?
+                 As part of migrating to Terraform Cloud, Terraform can optionally copy your
+                 current workspace state to the configured Terraform Cloud workspace.
+
+                 Answer "yes" to copy the latest state snapshot to the configured
+                 Terraform Cloud workspace.
+
+                 Answer "no" to ignore the existing state and just activate the configured
+                 Terraform Cloud workspace with its existing state, if any.
+
+                 Should Terraform migrate your existing state?
+
+                 Enter a value:
+   yes
+   Initializing provider plugins...
+               - Reusing previous version of hashicorp/random from the dependency lock file
+   - Using previously-installed hashicorp/random v3.4.3
+   Terraform Cloud has been successfully initialized!
    ```
-
-5. Navigate to `cdktf.out/stacks/hello-terraform` and run `terraform init`. CDKTF prints the following output:
-
-   ```
-   Initializing the backend...
-   Do you want to copy existing state to the new backend?
-     Pre-existing state was found while migrating the previous "local" backend to the
-     newly configured "remote" backend. No existing state was found in the newly
-     configured "remote" backend. Do you want to copy this state to the new "remote"
-     backend? Enter "yes" to copy and "no" to start with an empty state.
-
-     Enter a value: yes
-
-     Successfully configured the backend "remote"! Terraform will automatically
-     use this backend unless the backend configuration changes.
-
-     Initializing provider plugins...
-
-     .....
-     Terraform has been successfully initialized!
-
-     You may now begin working with Terraform. Try running "terraform plan" to see
-     any changes that are required for your infrastructure. All Terraform commands
-     should now work.
-
-     If you ever set or change modules or backend configuration for Terraform,
-     other cdktf cli commands will detect the change and prompt you to rerun this command if necessary.
-
-   ```
-
-6. Run`cdktf diff` in the root `hello-terraform` directory to validate state migration. CDKTF prints the following output:
-
-   ```bash
-   Stack: hello-terraform
-
-   Diff: 0 to create, 0 to update, 0 to delete.
-
-   ```
-
-   With the remote backend type in this example, a diff would indicate that the state was not migrated properly. There are no changes to the stack, which means the migration was successful. Consult the documentation for the remote backend you are using to understand how to validate state migration.
 
 # Supported Backends
 

